File: DatabindingWithNewsApi/app/src/main/java/com/enpassio/databindingwithnewsapi/ui/ArticleListFragment.java
package com.enpassio.databindingwithnewsapi.ui;

import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.enpassio.databindingwithnewsapi.R;
import com.enpassio.databindingwithnewsapi.databinding.NewsListBinding;
import com.enpassio.databindingwithnewsapi.model.Article;
import com.enpassio.databindingwithnewsapi.utils.UIState;
import com.enpassio.databindingwithnewsapi.viewmodel.MainViewModel;

public class ArticleListFragment extends Fragment implements NewsAdapter.ArticleClickListener{

    private MainViewModel mViewModel;
    private NewsAdapter mAdapter;
    private static final String TAG = "ArticleListFragment";
    private NewsListBinding binding;

    public ArticleListFragment() {
        setRetainInstance(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(
                inflater, R.layout.app_bar_main, container, false);

        //Set adapter, divider and default animator to the recycler view
        mAdapter = new NewsAdapter(this);
        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(requireActivity(),
                LinearLayoutManager.VERTICAL);
        binding.included.newsRecyclerView.addItemDecoration(dividerItemDecoration);
        binding.included.newsRecyclerView.setItemAnimator(new DefaultItemAnimator());
        binding.included.newsRecyclerView.setAdapter(mAdapter);

        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        //Get the view model instance and pass it to the binding implementation
        mViewModel = ViewModelProviders.of(requireActivity()).get(MainViewModel.class);
        binding.included.setViewModel(mViewModel);

        //Verify the connection and start loading from the api
        mViewModel.checkConnectionAndStartLoading();

        mViewModel.getConnectionStatus().observe(this, isConnected -> {
            /*If network is connected, set as "loading" until data arrives. If there
            is no connection, remove loading indicator and show no network image instead*/
            //If there is no network, show a snack bar to warn the user.
            if (!isConnected) {
                mViewModel.uiState.set(UIState.NETWORK_ERROR);
                showSnack();
            } else {
                mViewModel.uiState.set(UIState.LOADING);
            }
        });

        //Claim the list from the view model and observe the results
        mViewModel.getArticleList().observe(this, articles -> {
            if (articles != null && !articles.isEmpty()) {
                /*When articles are received, hide the loading indicator
                and pass the articles to the adapter*/
                mViewModel.uiState.set(UIState.SUCCESS);
                mAdapter.setArticleList(articles);
                Log.d(TAG, "articles are received. list size: " + articles.size());
            }
        });
    }

    private void showSnack() {
        //Show a snack bar for warning about the network connection and prompt user to try again with a button
        Snackbar snackbar = Snackbar
                .make(binding.mainContent, R.string.no_network_connection, Snackbar.LENGTH_INDEFINITE)
                /*If user will click "Retry", we'll check the connection again,
                and fetch the news, if there is network this time. Otherwise, snack will be shown again.*/
                .setAction(R.string.retry, view -> mViewModel.checkConnectionAndStartLoading())
                //Set the color of action button
                .setActionTextColor(getResources().getColor(R.color.colorAccent));
        //Set the background color of the snack bar
        snackbar.getView().setBackgroundColor(getResources().getColor(R.color.colorPrimary));
        snackbar.show();
    }

    @Override
    public void onArticleClicked(Article chosenArticle) {
        /*When an article from the list is clicked, pass that article to the viewmodel
        as the selected article. Than launch details fragment. Note that you don't
        need to pass the article in the bundle, since the details fragment will get the
        selected item from the same view model. */
        mViewModel.setChosenArticle(chosenArticle);
        FragmentManager fm = getFragmentManager();
        if (fm != null) {
            fm.beginTransaction()
                    .replace(R.id.fragment_holder, new ArticleDetailsFragment())
                    .addToBackStack(null)
                    .commit();
        }
    }
}

File: DatabindingWithNewsApiKotlin/app/src/main/java/com/enpassio/databindingwithnewsapikotlin/ui/ArticleListFragment.kt
package com.enpassio.databindingwithnewsapikotlin.ui


import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.transaction
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.enpassio.databindingwithnewsapikotlin.R
import com.enpassio.databindingwithnewsapikotlin.databinding.NewsListBinding
import com.enpassio.databindingwithnewsapikotlin.model.Article
import com.enpassio.databindingwithnewsapikotlin.model.UIState
import org.jetbrains.anko.design.indefiniteSnackbar


class ArticleListFragment : Fragment(), NewsAdapter.ArticleClickListener {

    private val mViewModel: MainViewModel by lazy {
        ViewModelProviders.of(requireActivity()).get(MainViewModel::class.java)
    }

    private lateinit var mAdapter: NewsAdapter
    private lateinit var binding: NewsListBinding

    init {
        retainInstance = true
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = DataBindingUtil.inflate(
            inflater, R.layout.app_bar_main, container, false
        )

        //Set adapter, divider and default animator to the recycler view
        mAdapter = NewsAdapter(this)
        val dividerItemDecoration = DividerItemDecoration(
            requireActivity(),
            LinearLayoutManager.VERTICAL
        )

        with(binding.included.newsRecyclerView) {
            addItemDecoration(dividerItemDecoration)
            itemAnimator = DefaultItemAnimator()
            adapter = mAdapter
        }

        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        //Get the view model instance and pass it to the binding implementation
        binding.included.uiState = mViewModel.uiState

        //When using livedata inside binding implementation, we should specify the lifecycle owner
        binding.lifecycleOwner = this.viewLifecycleOwner

        //Claim the list from the view model and observe the results
        mViewModel.articleList?.observe(this, Observer { articles ->
            if (!articles.isNullOrEmpty()) {
                /*When articles are received, pass the articles to the adapter
                And change uiState to SUCCESS. This will hide the loading indicator and show the list.*/
                mViewModel.setUiState(UIState.SUCCESS)
                mAdapter.articleList = articles
                Log.d(TAG, "articles are received. list size: " + articles.size)
            }
        })

        mViewModel.showSnack.observe(this, Observer { shouldShowSnack ->
            if (shouldShowSnack == true) {
                //Show a snack bar for warning about the network connection and prompt user to try again with a button
                binding.mainContent.indefiniteSnackbar(R.string.no_network_connection, R.string.retry) {
                    mViewModel.checkConnectionAndStartLoading()
                }
            }
        })
    }

    override fun onArticleClicked(chosenArticle: Article) {
        /*When an article from the list is clicked, pass that article to the viewmodel
        as the selected article. Than launch details fragment. Note that you don't
        need to pass the article in the bundle, since the details fragment will get the
        selected item from the same view model. */
        mViewModel.chosenArticle = chosenArticle
        fragmentManager?.transaction {
            replace(R.id.fragment_holder, ArticleDetailsFragment())
            addToBackStack(null)
        }
    }

    companion object {
        private const val TAG = "ArticleListFragment"
    }
}